lib/WUI/wui.cpp

```diff
@@ -0,0 +1,620 @@
+        netdev_status_t status = NETDEV_NETIF_DOWN;
+        with_iface(netdev_id, [&](netif &iface, NetworkState &instance) {
+            if (netif_is_link_up(&iface)) {
+                status = instance.netif_link(netdev_id) ? NETDEV_NETIF_UP : NETDEV_UNLINKED;
+            }
+        });
+        return status;
+    }
+    static uint32_t get_active() {
+        NetworkState *state = instance;
+        // TODO: Lock!
+        return state ? state->active : NETDEV_NODEV_ID;
+    }
+};
+
+std::atomic<NetworkState *> NetworkState::instance = nullptr;
+
+}
+
+void start_network_task() {
+    NetworkState::run_task();
+}
+
+const char *wui_get_api_key() {
+    return variant8_get_pch(prusa_link_api_key);
+}
+
+void notify_esp_data() {
+    NetworkState::notify(NetworkState::NetworkAction::EspData);
+}
+
+void notify_ethernet_data() {
+    NetworkState::notify(NetworkState::NetworkAction::EthData);
+}
+
+void netdev_get_ipv4_addresses(uint32_t netdev_id, lan_t *config) {
+    NetworkState::get_addresses(netdev_id, config);
+}
+
+void netdev_get_MAC_address(uint32_t netdev_id, uint8_t mac[OTP_MAC_ADDRESS_SIZE]) {
+    NetworkState::get_mac(netdev_id, mac);
+}
+
+const char *netdev_get_hostname(uint32_t netdev_id) {
+    return NetworkState::get_hostname(netdev_id);
+}
+
+netdev_status_t netdev_get_status(uint32_t netdev_id) {
+    return NetworkState::get_status(netdev_id);
+}
+
+uint32_t netdev_get_active_id() {
+    return NetworkState::get_active();
+}
+
+void notify_reconfigure() {
+    NetworkState::notify(NetworkState::NetworkAction::Reconfigure);
+}
+
+void netdev_set_active_id(uint32_t netdev_id) {
+    assert(netdev_id < NETDEV_COUNT);
+
+    eeprom_set_var(EEVAR_ACTIVE_NETDEV, variant8_ui8((uint8_t)(netdev_id & 0xFF)));
+
+    notify_reconfigure();
+}
+
+namespace {
+
+template <class F>
+void modify_flag(uint32_t netdev_id, F &&f) {
+    eevar_id var = EEVAR_LAN_FLAG;
+    switch (netdev_id) {
+    case NETDEV_ETH_ID:
+        var = EEVAR_LAN_FLAG;
+        break;
+    case NETDEV_ESP_ID:
+        var = EEVAR_WIFI_FLAG;
+        break;
+    default:
+        assert(0);
+    }
+
+    // Read it from the EEPROM, not from the state. For two reasons:
+    // * While it likely can't happen, it's unclear what should happen if the
+    //   state wasn't initiated yet.
+    // * The config in the state is delayed (reloaded in reconfigure), we want
+    //   as fresh value as possible. This still leaves the possibility of a
+    //   race condition (two threads messing with the same variable), but that
+    //   is unlikely.
+    const uint8_t old = variant8_get_ui8(eeprom_get_var(var));
+    uint8_t flag = f(old);
+    if (old != flag) {
+        eeprom_set_var(var, variant8_ui8(flag));
+        notify_reconfigure();
+    }
+}
+
+}
+
+void netdev_set_static(uint32_t netdev_id) {
+    modify_flag(netdev_id, [](uint8_t flag) -> uint8_t {
+        CHANGE_FLAG_TO_STATIC(flag);
+        TURN_FLAG_ON(flag);
```

Comment:
I don't like these macros but let's keep them.

Reply:
I don't like them either. But I didn't want to fix everything I don't like about the code in a single commit, the review would be even harder than now 😈